File: app/TumbleKit/Views/PrintView.swift
import SwiftUI
import UIKit

/// A single instant print, mounted on cream stock — the exact treatment from
/// the site's `DrawerMockup`: the photograph, a warm aged grade that grows
/// with age, film grain, a vignette, and a soft sheen. Also renders the blank,
/// face-down state for an undeveloped shot.
///
/// `developProgress` (0…1) drives the shake-to-develop look: it starts washed
/// out and desaturated and settles into full color.
public struct PrintView: View {
    private let image: UIImage?
    private let isDeveloped: Bool
    private let developProgress: Double
    private let age: Double
    private let caption: String?
    private let width: CGFloat

    public init(
        image: UIImage?,
        isDeveloped: Bool,
        developProgress: Double = 1,
        age: Double = 0,
        caption: String? = nil,
        width: CGFloat = 200
    ) {
        self.image = image
        self.isDeveloped = isDeveloped
        self.developProgress = developProgress
        self.age = age
        self.caption = caption
        self.width = width
    }

    public var body: some View {
        VStack(spacing: 0) {
            photo
                .aspectRatio(1, contentMode: .fit)
                .clipShape(RoundedRectangle(cornerRadius: width * 0.01))
                .overlay(
                    RoundedRectangle(cornerRadius: width * 0.01)
                        .strokeBorder(.black.opacity(0.15), lineWidth: 0.5)
                )

            if let caption, isDeveloped {
                Text(caption)
                    .font(Typography.display(width * 0.052, weight: .regular))
                    .italic()
                    .foregroundStyle(Palette.ink.opacity(0.7))
                    .padding(.top, width * 0.06)
            }
        }
        .padding(width * 0.06)
        .padding(.bottom, width * 0.09)
        .frame(width: width)
        .background(Palette.printStock)
        .clipShape(RoundedRectangle(cornerRadius: width * 0.025))
        .shadow(color: .black.opacity(0.5), radius: width * 0.06, x: 0, y: width * 0.05)
    }

    @ViewBuilder private var photo: some View {
        if isDeveloped || developProgress > 0 {
            developedPhoto
        } else {
            // Blank, face-down: an undeveloped shot you haven't shaken yet.
            ZStack {
                LinearGradient(
                    colors: [Color(hex: 0xE8DFCC), Color(hex: 0xD8CDB4)],
                    startPoint: .top, endPoint: .bottom
                )
                Image(systemName: "hand.draw")
                    .font(.system(size: width * 0.12, weight: .light))
                    .foregroundStyle(Palette.ink.opacity(0.25))
            }
        }
    }

    private var developedPhoto: some View {
        ZStack {
            sceneLayer
            agedGrade
            GrainOverlay(opacity: 0.4)
            vignette
            sheen
        }
        // Develop transition: washed out + desaturated at first, settling in.
        .saturation(developProgress)
        .overlay(Color.white.opacity((1 - developProgress) * 0.65))
        .brightness((1 - developProgress) * 0.18)
    }

    // Warm aged grade — the site formula: warm over cool, both scaling with age.
    private var agedGrade: some View {
        LinearGradient(
            colors: [
                Color(hex: 0xD6965A, opacity: 0.12 + age * 0.24),
                Color(hex: 0x78463C, opacity: 0.06 + age * 0.16),
            ],
            startPoint: .topLeading, endPoint: .bottomTrailing
        )
        .blendMode(.multiply)
    }

    private var vignette: some View {
        RadialGradient(
            colors: [.clear, Color(hex: 0x1C1012, opacity: 0.42)],
            center: .init(x: 0.5, y: 0.44),
            startRadius: width * 0.3,
            endRadius: width * 0.72
        )
    }

    private var sheen: some View {
        LinearGradient(
            colors: [.white.opacity(0.16), .clear],
            startPoint: .topLeading, endPoint: .init(x: 0.34, y: 0.34)
        )
    }

    @ViewBuilder private var sceneLayer: some View {
        if let image {
            Image(uiImage: image)
                .resizable()
                .scaledToFill()
        } else {
            Color(hex: 0x2A3A49)
        }
    }
}

File: android/app/src/main/java/com/tumble/ui/components/PrintView.kt
package com.tumble.ui.components

import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.RenderEffect
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Draw
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asComposeRenderEffect
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.tumble.ui.theme.Palette
import com.tumble.ui.theme.TumbleType
import com.tumble.ui.theme.grain

/**
 * A single instant print, mounted on cream stock — the exact treatment from the
 * site's `DrawerMockup`: the photograph, a warm aged grade that grows with age,
 * film grain, a vignette, and a soft sheen. Also renders the blank, face-down
 * state for an undeveloped shot.
 *
 * [developProgress] (0…1) drives the shake-to-develop look: it starts washed out
 * and desaturated and settles into full color. Ported from
 * `app/TumbleKit/Views/PrintView.swift`.
 */
@Composable
fun PrintView(
    image: ImageBitmap?,
    isDeveloped: Boolean,
    modifier: Modifier = Modifier,
    developProgress: Float = 1f,
    age: Float = 0f,
    caption: String? = null,
    width: Dp = 200.dp,
) {
    val pad = width * 0.06f
    Column(
        modifier = modifier
            .width(width)
            .shadow(width * 0.06f, RoundedCornerShape(width * 0.025f))
            .clip(RoundedCornerShape(width * 0.025f))
            .background(Palette.printStock)
            .padding(PaddingValues(start = pad, top = pad, end = pad, bottom = width * 0.09f)),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        val photoCorner = RoundedCornerShape(width * 0.01f)
        androidx.compose.foundation.layout.Box(
            modifier = Modifier
                .fillMaxWidth()
                .aspectRatio(1f)
                .clip(photoCorner)
                .border(0.5.dp, Color.Black.copy(alpha = 0.15f), photoCorner),
        ) {
            if (isDeveloped || developProgress > 0f) {
                DevelopedPhoto(image = image, developProgress = developProgress, age = age)
            } else {
                UndevelopedFace(width = width)
            }
        }

        if (caption != null && isDeveloped) {
            Text(
                text = caption,
                style = TumbleType.display((width.value * 0.052f).toInt().coerceAtLeast(8), FontWeight.Normal)
                    .copy(fontStyle = FontStyle.Italic, color = Palette.ink.copy(alpha = 0.7f)),
                modifier = Modifier.padding(top = width * 0.06f),
            )
        }
    }
}

@Composable
private fun DevelopedPhoto(image: ImageBitmap?, developProgress: Float, age: Float) {
    val p = developProgress.coerceIn(0f, 1f)
    androidx.compose.foundation.layout.Box(Modifier.fillMaxSize()) {
        // The develop transition desaturates the whole print at first (saturation
        // ramps with progress) — applied as a layer RenderEffect (API 31+).
        androidx.compose.foundation.layout.Box(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer {
                    clip = true
                    renderEffect = saturationEffect(p)
                }
                .drawWithContent {
                    drawContent() // the scene image / fallback
                    // Warm aged grade — warm over cool, both scaling with age.
                    drawRect(
                        brush = Brush.linearGradient(
                            colors = listOf(
                                Color(0xFFD6965A).copy(alpha = 0.12f + age * 0.24f),
                                Color(0xFF78463C).copy(alpha = 0.06f + age * 0.16f),
                            ),
                            start = Offset.Zero,
                            end = Offset(size.width, size.height),
                        ),
                        blendMode = BlendMode.Multiply,
                    )
                    // Film grain.
                    drawRect(brush = com.tumble.ui.theme.Grain.brush, alpha = 0.4f, blendMode = BlendMode.Overlay)
                    // Vignette — clear center, dark edges (center biased upward).
                    drawRect(
                        brush = Brush.radialGradient(
                            0.42f to Color.Transparent,
                            1f to Color(0xFF1C1012).copy(alpha = 0.42f),
                            center = Offset(0.5f * size.width, 0.44f * size.height),
                            radius = 0.72f * size.width,
                        ),
                    )
                    // Sheen — glossy highlight top-left.
                    drawRect(
                        brush = Brush.linearGradient(
                            colors = listOf(Color.White.copy(alpha = 0.16f), Color.Transparent),
                            start = Offset.Zero,
                            end = Offset(0.34f * size.width, 0.34f * size.height),
                        ),
                    )
                },
        ) {
            if (image != null) {
                androidx.compose.foundation.Image(
                    bitmap = image,
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.fillMaxSize(),
                )
            } else {
                androidx.compose.foundation.layout.Box(
                    Modifier.fillMaxSize().background(Color(0xFF2A3A49)),
                )
            }
        }

        // Washed-out overlay + brightness lift that fade as the print develops.
        androidx.compose.foundation.layout.Box(
            Modifier.fillMaxSize().background(Color.White.copy(alpha = (1f - p) * 0.65f)),
        )
        androidx.compose.foundation.layout.Box(
            Modifier.fillMaxSize().background(Color.White.copy(alpha = (1f - p) * 0.18f)),
        )
    }
}

@Composable
private fun UndevelopedFace(width: Dp) {
    androidx.compose.foundation.layout.Box(
        modifier = Modifier
            .fillMaxSize()
            .background(
                Brush.verticalGradient(listOf(Color(0xFFE8DFCC), Color(0xFFD8CDB4))),
            ),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = Icons.Outlined.Draw,
            contentDescription = null,
            tint = Palette.ink.copy(alpha = 0.25f),
            modifier = Modifier.width(width * 0.12f),
        )
    }
}

/** Saturation color-matrix as a Compose [androidx.compose.ui.graphics.RenderEffect]. */
private fun saturationEffect(progress: Float): androidx.compose.ui.graphics.RenderEffect {
    val matrix = ColorMatrix().apply { setSaturation(progress) }
    return RenderEffect
        .createColorFilterEffect(ColorMatrixColorFilter(matrix))
        .asComposeRenderEffect()
}
